src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,244 @@
+
+/-!  #  Max-degree and min-degree of an `add_monoid_algebra`
+
+Let `R` be a semiring and let `A` be a `semilattice_sup`.
+
+For an element `f : add_monoid_algebra R A`, this file defines
+* `add_monoid_algebra.max_degree`: the max-degree taking values in `with_bot A`,
+* `add_monoid_algebra.min_degree`: the min-degree taking values in `with_top A`.
+If the grading type `A` is a linearly ordered additive monoid, then these two notions of degree
+coincide with the standard one:
+* the max-degree is the maximum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊥`, if `f = 0`;
+* the min-degree is the minimum of the exponents of the monomials that appear with non-zero
+  coefficient in `f`, or `⊤`, if `f = 0`.
+
+The main results are
+* `add_monoid_algebra.max_degree_mul_le`:
+  the max-degree of a product is at most the sum of the max-degrees,
+* `add_monoid_algebra.le_min_degree_mul`:
+  the min-degree of a product is at least the sum of the min-degrees,
+* `add_monoid_algebra.max_degree_add_le`:
+  the max-degree of a sum is at most the sup of the max-degrees,
+* `le_min_degree_add`: the min-degree of a sum is at least the inf of the min-degrees.
+
+## Implementation notes
+
+The current plan is to state and prove lemmas about `finset.sup (finsupp.support f) D` with a
+"generic" degree/weight function `D` from the grading Type `A` to a somewhat ordered Type `B`.
+
+Next, the general lemmas get specialized twice:
+* once for `max_degree` (essentially a simple application) and
+* once for `min_degree` (a simple application, via `order_dual`).
+These final lemmas are the ones that likely get used the most.  The generic lemmas about
+`finset.support.sup` may not be used directly much outside of this file.
+
+To see this in action, you can look at the triple
+`(sup_support_mul_le, max_degree_mul_le, le_min_degree_mul)`.
+-/
+
+variables {R A B ι : Type*}
+
+namespace add_monoid_algebra
+open_locale classical big_operators
+
+/-! ### Results about the `finset.sup` and `finset.inf` of `finsupp.support` -/
+
+section general_results_assuming_semilattice_sup
+
+section semiring
+
+variables [semiring R]
+
+lemma sup_support_add_le [semilattice_sup B] [order_bot B]
+  (D : A → B) (f g : add_monoid_algebra R A) :
+  (f + g).support.sup D ≤ (f.support.sup D) ⊔ (g.support.sup D) :=
+begin
+  refine (finset.sup_le (λ d ds, _)),
+  cases finset.mem_union.mp (finsupp.support_add ds) with df dg,
+  { exact (finset.le_sup df).trans le_sup_left },
+  { exact (finset.le_sup dg).trans le_sup_right }
+end
+
+lemma le_inf_degree_add [semilattice_inf B] [order_top B]
+  (D : A → B) (f g : add_monoid_algebra R A) :
+  f.support.inf D ⊓ g.support.inf D ≤ (f + g).support.inf D :=
+sup_support_add_le (λ a : A, order_dual.to_dual (D a)) f g
+
+section add_only
+
+variables [has_add A] [has_add B]
+
+lemma sup_support_mul_le [semilattice_sup B] [order_bot B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
```

Comment:
This section would probably benefit from a `variables` statement, and two different variables for `order_bot _` and `order_top B`. In the absence of any inventive ideas for naming, that could just be `B_1` and `B_2`

Reply:
I prefer the two variables and side-by-side lemmas.  Since the lemmas would be updated in pairs, I think that it is easier to keep them in sync.  I will probably go with `T` for the Type `with_top` and `B` for the Type `with_bot`.